velox/common/memory/MemoryPool.h

```diff
@@ -890,6 +890,142 @@ class MemoryPoolImpl : public MemoryPool {
   // The number of internal memory reservation collisions caused by concurrent
   // memory reservation requests.
   std::atomic<uint64_t> numCollisions_{0};
+
+#define MEMORY_DEBUG
+#ifdef MEMORY_DEBUG
+#define RECORD_ALLOC_DEBUG(...) recordAllocDbg(__VA_ARGS__)
+#define RECORD_FREE_DEBUG(...) recordFreeDbg(__VA_ARGS__)
+#define LEAK_CHECK_DEBUG() leakCheckDbg()
+
+  std::mutex dbgAllocMutex_;
+  // Map from address to {alloc size, call stack}
+  std::unordered_map<uint64_t, std::pair<uint64_t, std::string>>
+      addrToAllocDetail_;
+
+  /// Recording on every allocation is very expensive and normally will result
+  /// in CPU saturation. Modify this method while debugging to limit the number
+  /// of times that the allocations are recorded. 'isAlloc' will be true at
+  /// allocation sites, false at free sites. You want to specify the condition
+  /// so that it is guaranteed that each alloc recording will have a
+  /// corresponding free recording called. In other words, always make sure the
+  /// condition at free sites is less strict than the condition at alloc sites.
+  inline bool allocDbgConditionMet(bool isAlloc) {
+    return true;
+  }
+
+  /// Debug method that is only enabled when 'MEMORY_DBG' macro is defined. This
+  /// should be called after actual allocation.
+  inline void recordAllocDbg(const void* addr, uint64_t size) {
+    if (!allocDbgConditionMet(true)) {
+      return;
+    }
+    const auto stackTrace = process::StackTrace().toString();
+    std::lock_guard<std::mutex> l(dbgAllocMutex_);
+    addrToAllocDetail_.emplace(
+        reinterpret_cast<uint64_t>(addr), std::pair{size, stackTrace});
+  }
+
+  /// Debug method that is only enabled when 'MEMORY_DBG' macro is defined. This
+  /// should be called after actual allocation.
+  inline void recordAllocDbg(const Allocation& allocation) {
+    if (!allocDbgConditionMet(true) || allocation.empty()) {
+      return;
+    }
+    recordAllocDbg(allocation.runAt(0).data(), allocation.byteSize());
+  }
+
+  /// Debug method that is only enabled when 'MEMORY_DBG' macro is defined. This
+  /// should be called after actual allocation.
+  inline void recordAllocDbg(const ContiguousAllocation& allocation) {
+    if (!allocDbgConditionMet(true) || allocation.empty()) {
+      return;
+    }
+    recordAllocDbg(allocation.data(), allocation.size());
+  }
+
+  /// Debug method that is only enabled when 'MEMORY_DBG' macro is defined. This
+  /// debug method should be called strictly before the actual free happens.
+  /// This is so to avoid the following race condition example if called after:
+  /// - Thread 1 calls allocate for 128 bytes.
```

Comment:
We usually don't put such kind of detailed comments like the race condition between threads in header file or even in code implementation. This is more like a bug fix description instead of code comment. Probably better to move this to PR description if it is worth tracking.